applications/serial_lte_modem/src/lwm2m_carrier/slm_at_carrier_cfg.c

```diff
@@ -0,0 +1,799 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+#include <stdio.h>
+#include <string.h>
+#include <zephyr/logging/log.h>
+#include <lwm2m_carrier.h>
+#include <lwm2m_settings.h>
+#include "slm_at_host.h"
+#include "slm_settings.h"
+#include "slm_util.h"
+
+LOG_MODULE_REGISTER(slm_carrier_cfg, CONFIG_SLM_LOG_LEVEL);
+
+/**@brief LwM2M Carrier Configuration operations. */
+enum slm_carrier_cfg_operation {
+	/* Enable Automatic Startup */
+	CARRIER_CFG_OP_AUTO_STARTUP,
+	/* General Custom Configuration */
+	CARRIER_CFG_OP_CONFIG_ENABLE,
+	CARRIER_CFG_OP_CONFIG_APN,
+	CARRIER_CFG_OP_CONFIG_BOOTSTRAP_FROM_SMARTCARD,
+	CARRIER_CFG_OP_CONFIG_CARRIERS,
+	CARRIER_CFG_OP_CONFIG_COAP_CON_INTERVAL,
+	CARRIER_CFG_OP_CONFIG_LGU_DEVICE_SERIAL_NO_TYPE,
+	CARRIER_CFG_OP_CONFIG_LGU_SERVICE_CODE,
+	CARRIER_CFG_OP_CONFIG_PDN_TYPE,
+	CARRIER_CFG_OP_CONFIG_SESSION_IDLE_TIMEOUT,
+	/* Device Custom Configuration */
+	CARRIER_CFG_OP_DEVICE_ENABLE,
+	CARRIER_CFG_OP_DEVICE_DEVICE_TYPE,
+	CARRIER_CFG_OP_DEVICE_HW_VERSION,
+	CARRIER_CFG_OP_DEVICE_MANUFACTURER,
+	CARRIER_CFG_OP_DEVICE_MODEL_NUMBER,
+	CARRIER_CFG_OP_DEVICE_SW_VERSION,
+	/* Server Custom Configuration */
+	CARRIER_CFG_OP_SERVER_ENABLE,
+	CARRIER_CFG_OP_SERVER_BINDING,
+	CARRIER_CFG_OP_SERVER_IS_BOOTSTRAP,
+	CARRIER_CFG_OP_SERVER_LIFETIME,
+	CARRIER_CFG_OP_SERVER_SEC_TAG,
+	CARRIER_CFG_OP_SERVER_URI,
+	/* Count */
+	CARRIER_CFG_OP_MAX
+};
+
+/** forward declaration of configuration cmd handlers **/
+static int do_cfg_apn(void);
+static int do_cfg_auto_startup(void);
+static int do_cfg_bootstrap_from_smartcard(void);
+static int do_cfg_carriers(void);
+static int do_cfg_coap_con_interval(void);
+static int do_cfg_config_enable(void);
+static int do_cfg_device_enable(void);
+static int do_cfg_device_type(void);
+static int do_cfg_hardware_version(void);
+static int do_cfg_manufacturer(void);
+static int do_cfg_model_number(void);
+static int do_cfg_session_idle_timeout(void);
+static int do_cfg_software_version(void);
+static int do_cfg_device_serial_no_type(void);
+static int do_cfg_service_code(void);
+static int do_cfg_pdn_type(void);
+static int do_cfg_binding(void);
+static int do_cfg_server_enable(void);
+static int do_cfg_is_bootstrap(void);
+static int do_cfg_lifetime(void);
+static int do_cfg_sec_tag(void);
+static int do_cfg_uri(void);
+
+struct carrier_cfg_op_list {
+	uint8_t op_code;
+	char *op_str;
+	int (*handler)(void);
+};
+
+/**@brief SLM Carrier Configuration AT Command list type. */
+static struct carrier_cfg_op_list cfg_op_list[CARRIER_CFG_OP_MAX] = {
+	{CARRIER_CFG_OP_AUTO_STARTUP, "auto_startup", do_cfg_auto_startup},
```

Comment:
The `op_code` values do not seem to be used anywhere.
Also, for your information, you can explicitly give indices while assigning the array.
This would give:
```suggestion
	[CARRIER_CFG_OP_AUTO_STARTUP] = {"auto_startup", do_cfg_auto_startup},
```

Reply:
It is indeed not used. For consistency, I followed the same format as in `slm_at_carrier.c`, which in turn followed `slm_at_ftp.c` if I'm not mistaken. Perhaps this cleanup could be handled in a follow-up PR to address all 3 files at once? I'd be happy to open one myself. If not, I can adapt only the new code here.

Reply:
Ah okay. Indeed in `slm_at_ftp.c` the same is done. SLM code is not always the best example to follow. :smiley:
And I get that you did the same for consistency.
I agree that these could be addressed in a dedicated PR.